client/txn.go

```diff
@@ -300,18 +301,33 @@ func (txn *Txn) CommitInBatch(b *Batch) error {
 
 // Commit sends an EndTransactionRequest with Commit=true.
 func (txn *Txn) Commit() error {
-	if txn.Proto.Writing {
-		return txn.sendEndTxnCall(true /* commit */)
+	err := txn.sendEndTxnCall(true /* commit */)
+	if err == nil {
+		return nil
 	}
-	return nil
+	if !txn.execing {
+		// If this transaction is not part of the auto-retry loop we need to
+		// cleanup the transaction on retryable commit failures.
+		if restartErr, ok := err.(proto.TransactionRestartError); ok {
+			if restartErr.CanRestartTransaction() != proto.TransactionRestart_ABORT {
```

Comment:
by the way, why the special case here? the exec loop does actually send a rollback on `TransactionRestart_ABORT`